### PR TITLE
Upgrade to v0.5.0 wasm3 API

### PR DESCRIPTION
**Pull request by jimpo**
Addresses #24 

This is a big change to support the v0.5.0 wasm3 release. I take the approach of only using the public API in `wasm3.h` and `m3_api_wasi.h` (if the `wasi` feature is enabled). This means I drop the following functions:

- `Function::import_module_name` (The name of the import module of this function)
- `Module::function` (Looks up a function by its index in this module)
- `Runtime::find_module` (Searches for a module with the given name in the runtime's loaded modules)
- `Runtime::modules` (Returns an iterator over the runtime's loaded modules)
- `Runtime::resize_memory` (Resizes the number of allocatable pages to num_pages)
- `Runtime::stack and stack_mut` (Returns the stack of this runtime)

Aside from this I try to keep full compatibility with the current API.

### CODE REVIEW

**Comment by Veykril** [comment]
Will take a look later this week(maybe tomorrow) but in general the crate was using a lot of internal things because wasm3 itself didn't really expose a nice api at the time. So getting rid of the internal dependence is definitely a plus point assuming most of the things can now be accessed via the public api.

File: src/function.rs [hunk 1]
```diff
@@ -25,24 +24,15 @@ impl<'cc> CallContext<'cc> {
...
-    unsafe fn mallocated(&self) -> *mut ffi::M3MemoryHeader {
-        self.runtime.as_ref().memory.mallocated
-    }
-
     /// Returns the raw memory of the runtime associated with this context.
     ///
     /// # Safety
     ///
     /// The returned pointer may get invalidated when wasm function objects are called due to reallocations.
     pub unsafe fn memory(&self) -> *const [u8] {
```
Comment (Veykril) [inline src/function.rs:32]: Can you remove the `unsafe` qualifier of this function while we are touching this already? It returns a raw pointer so calling it is actually safe, using the pointer is the unsafe operation here. Same for `memory_mut` below

File: src/function.rs [hunk 2]
```diff
@@ -116,77 +107,43 @@ where
...
+        if !Self::validate_sig(raw) {
+            return Err(Error::InvalidFunctionSignature);
+        }
+        Ok(Function {
             raw,
             rt,
             _pd: PhantomData,
-        };
-        // make sure the function is compiled
-        this.compile()
```
Comment (Veykril) [inline src/function.rs:143]: Are functions compiled by default now?

File: src/function.rs [hunk 3]
```diff
@@ -116,77 +107,43 @@ where
...
-        }
+    fn validate_sig(func: NNM3Function) -> bool {
+        let num_args = unsafe { ffi::m3_GetArgCount(func.as_ptr()) };
+        let num_rets = unsafe { ffi::m3_GetRetCount(func.as_ptr()) };
+        let args = (0..num_args)
+            .map(|i| unsafe { ffi::m3_GetArgType(func.as_ptr(), i) })
+            .collect::<Vec<_>>();
+        let rets = (0..num_rets)
+            .map(|i| unsafe { ffi::m3_GetRetType(func.as_ptr(), i) })
+            .collect::<Vec<_>>();
```
Comment (Veykril) [inline src/function.rs:118]: Not a fan of these collects, would be nice if we could avoid these somehow

File: src/module.rs [hunk 4]
```diff
@@ -38,8 +37,8 @@ impl ParsedModule {
         self.raw
     }
 
-    pub(crate) fn take_data(self) -> Box<[u8]> {
-        let res = unsafe { ptr::read(&self.data) };
+    pub(crate) fn take_data(mut self) -> Box<[u8]> {
+        let res = mem::replace(&mut self.data, <Box<[u8]>>::default());
```
Comment (Veykril) [inline src/module.rs:41]: What's the reason for this change? This leaks memory now

File: src/module.rs [hunk 5]
```diff
@@ -99,9 +99,20 @@ impl<'rt> Module<'rt> {
         Args: crate::WasmArgs,
         Ret: crate::WasmType,
     {
-        let func = self.find_import_function(module_name, function_name)?;
-        Function::<'_, Args, Ret>::validate_sig(func)
-            .and_then(|_| unsafe { self.link_func_impl(func, f) })
+        let module_name_cstr = str_to_cstr_owned(module_name);
+        let function_name_cstr = str_to_cstr_owned(function_name);
+        let signature = function_signature::<Args, Ret>();
```
Comment (Veykril) [inline src/module.rs:104]: Having to realloc these for null-termination is a shame but I guess we can't get around this

File: src/module.rs [hunk 6]
```diff
@@ -61,6 +60,7 @@ impl Drop for ParsedModule {
 pub struct Module<'rt> {
     raw: ffi::IM3Module,
     rt: &'rt Runtime,
+    name_cstr: Vec<cty::c_char>,
```
Comment (Veykril) [inline src/module.rs:63]: This is here to keep the name for `set_name` alive right? This is a problem as when this Module gets dropped we don't unload it from the runtime, so wasm3 has a dangling reference to the name then.

File: src/runtime.rs [hunk 7]
```diff
@@ -137,13 +99,9 @@ impl Runtime {
     ///
     /// The returned pointer may get invalidated when wasm function objects are called due to reallocations.
     pub unsafe fn memory(&self) -> *const [u8] {
```
Comment (Veykril) [inline src/runtime.rs:101]: Here and below the `unsafe` can go as well

File: src/ty.rs [hunk 8]
```diff
@@ -77,29 +29,35 @@ pub trait WasmArg: WasmType {}
...
     #[doc(hidden)]
-    unsafe fn push_on_stack(self, stack: *mut [ffi::m3slot_t]);
+    unsafe fn push_on_stack(self, stack: *mut u64);
     #[doc(hidden)]
     // required for closure linking
-    unsafe fn pop_from_stack(stack: *mut [ffi::m3slot_t]) -> Self;
+    unsafe fn pop_from_stack(stack: *mut u64) -> Self;
     #[doc(hidden)]
-    fn validate_types(types: &[u8]) -> bool;
+    fn validate_types(types: &[ffi::M3ValueType::Type]) -> bool;
```
Comment (Veykril) [inline src/ty.rs:37]:
```suggestion
    fn validate_types(types: impl Iterator<Item = ffi::M3ValueType::Type>) -> bool;
```
Doing this should allow us to get rid of the allocations annotated earlier

File: src/function.rs (hunk 2 above)
Comment (jimpo) [reply src/function.rs:143]: Yes, now the only way to get a function is `m3_FindFunction` which will compile

File: src/module.rs (hunk 5 above)
Comment (jimpo) [reply src/module.rs:104]: Yeah...

File: src/module.rs (hunk 4 above)
Comment (jimpo) [reply src/module.rs:41]: I wanted to get rid of the `unsafe`, and I figured that `<Box<[u8]>>::default()` wouldn't do any allocation because `[u8; 0]` is a zero-sized type.

Looking again though, I think this leaks anyway and did before because the `Environment` is not dropped and thus doesn't decrement the reference count.

The better way to to this would be to not implement `Drop` on `ParsedModule` but rather have it own a `DropModule(ffi::IM3Module)` type.

File: src/function.rs (hunk 1 above)
Comment (jimpo) [reply src/function.rs:32]: This can be handled in a separate issue/PR, but I think there should be a safer, non-pointer interface to memory. Either something that does runtime checking like `Cell` or if, and this would require a big API change, you need to get a `MemoryLease` from the `Runtime` to do anything that could modify memory, including `find_function` and `call`.

File: src/module.rs (hunk 6 above)
Comment (jimpo) [reply src/module.rs:63]: mmmm, you're right. I'm gonna drop `set_name` from this PR and open an issue to add as a follow-up.

**Comment by vshymanskyy** [comment]
@jimpo thanks for a valuable contribution!

File: src/function.rs (hunk 1 above)
Comment (Veykril) [reply src/function.rs]: Yes this should certainly have a proper API, this was just a quick way to expose the memory.

File: src/module.rs (hunk 4 above)
Comment (Veykril) [reply src/module.rs]: Right this does indeed leak the environment. So ye this should probably own a type that drops the module instead of being responsible for freeing the module itself as you've said, as then we can destructure `self` here.
